src/gst.c

```diff
@@ -0,0 +1,218 @@
+    pthread_cond_broadcast(&ctx->snapshot_cond);
+    pthread_mutex_unlock(&ctx->snapshot_mutex);
+    gst_buffer_unmap(buffer, &map);
+  }
+  return GST_PAD_PROBE_OK;
+}
+
+// updates stream parameters
+void gst_update_stream_params(gst_thread_ctx *ctx, int h264_bitrate, const char *rtp_host, int rtp_port) {
+  pthread_mutex_lock(&ctx->params_mutex);
+  // copy values into the gst_thread_ctx
+  ctx->h264_bitrate = h264_bitrate;
+  strncpy(ctx->rtp_host, rtp_host, sizeof(ctx->rtp_host));
+  ctx->rtp_port = rtp_port;
+  pthread_mutex_unlock(&ctx->params_mutex);
+}
+
+static gboolean gst_update_stream_params_cb(gpointer user_data) {
+  gst_thread_ctx *ctx = (gst_thread_ctx *)user_data;
+  pthread_mutex_lock(&ctx->params_mutex);
+  guint prev_h264_bitrate;
+  g_object_get(G_OBJECT(ctx->x264enc), "bitrate", &prev_h264_bitrate, NULL);
+  if (prev_h264_bitrate != ctx->h264_bitrate) {
+    INFO("gst_update_stream_params_cb(): bitrate: %dkbps", ctx->h264_bitrate);
+    g_object_set(G_OBJECT(ctx->x264enc), "bitrate", ctx->h264_bitrate, NULL);
+  }
+  gchar *prev_rtp_host;
+  gint prev_rtp_port;
+  g_object_get(G_OBJECT(ctx->udpsink), "host", &prev_rtp_host, "port", &prev_rtp_port, NULL);
+  if (strcmp(prev_rtp_host, ctx->rtp_host) != 0 || prev_rtp_port != ctx->rtp_port) {
+    INFO("gst_update_stream_params_cb(): RTP sink: %s:%d", ctx->rtp_host, ctx->rtp_port);
+    g_object_set(G_OBJECT(ctx->udpsink), "host", ctx->rtp_host, "port", (gint)ctx->rtp_port, NULL);
+  }
+  g_free(prev_rtp_host);
+  pthread_mutex_unlock(&ctx->params_mutex);
+  return G_SOURCE_CONTINUE;
+}
+
+static void *gst_thread(void *param) {
+  gst_thread_ctx *ctx = (gst_thread_ctx *)param;
+
+  // initialize gst_thread_ctx
+  pthread_mutex_init(&ctx->snapshot_mutex, NULL);
+  pthread_cond_init(&ctx->snapshot_cond, NULL);
+  ctx->snapshot_buf = NULL;
+  ctx->snapshot_buf_capacity = ctx->snapshot_size = 0;
+  ctx->current_frame = 0;
+  ctx->snapshot_frame = -1;
+  pthread_mutex_init(&ctx->params_mutex, NULL);
+  ctx->h264_bitrate = 256;
+  strcpy(ctx->rtp_host, "localhost");
+  ctx->rtp_port = 20000;
+
+  // initialize GStreamer/main loop
+  gst_init(NULL, NULL);
+  ctx->loop = g_main_loop_new(NULL, FALSE);
+
+  // create the stream pipeline
+  ctx->pipeline = gst_pipeline_new("default");
+  ctx->source = gst_element_factory_make("videotestsrc", "source");
+  ctx->source_filter = gst_element_factory_make ("capsfilter", "source-filter");
+  ctx->timeoverlay = gst_element_factory_make("timeoverlay", "overlay");
+  ctx->tee = gst_element_factory_make ("tee", "tee");
+  ctx->x264enc = gst_element_factory_make("x264enc", "x264-encoder");
+  g_object_set(G_OBJECT(ctx->x264enc), "bitrate", ctx->h264_bitrate, NULL);
+  ctx->enc_filter = gst_element_factory_make ("capsfilter", "enc-filter");
+  ctx->rtph264pay = gst_element_factory_make("rtph264pay", "rtph264");
+  g_object_set(G_OBJECT(ctx->rtph264pay), "config-interval", 1, NULL);
+  g_object_set(G_OBJECT(ctx->rtph264pay), "pt", 96, NULL);
```

Comment:
what's this param?

Reply:
"pt" here is RTP payload type (https://tools.ietf.org/html/rfc3551#section-6). 96 is what's currently configured on Janus gateways, could be anything in 96-127 range, I think